### PR TITLE
Add support for SSRF vulnerabilities

### PR DESCRIPTION
**Pull request by manuel-alvarez-alvarez**
# What Does This Do
Adds support for SSRF detection via jdk (URL/URI), okhttp (versions 2 and 3) and commons httpclient (versions 1 and 2)

# Motivation

# Additional Notes
Support for other frameworks like apache commons (4 and 5), spring rest-template will come later

### CODE REVIEW

File: dd-java-agent/instrumentation/java-net/src/main/java/datadog/trace/instrumentation/java/net/URLCallSite.java [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+  @CallSite.After("void java.net.URL.<init>(java.net.URL, java.lang.String)")
+  @CallSite.After(
+      "void java.net.URL.<init>(java.net.URL, java.lang.String, java.net.URLStreamHandler)")
+  public static URL afterCtor(
+      @CallSite.AllArguments final Object[] args, @CallSite.Return @Nonnull final URL result) {
+    if (args != null && args.length > 0) {
+      final PropagationModule module = InstrumentationBridge.PROPAGATION;
+      if (module != null) {
+        try {
+          module.taintIfAnyInputIsTainted(result, args);
```
Comment (manuel-alvarez-alvarez) [inline dd-java-agent/instrumentation/java-net/src/main/java/datadog/trace/instrumentation/java/net/URLCallSite.java:35]: I think it makes sense to build a different module aka `UrlModule` to deal with the propagation in URLs

File: dd-java-agent/agent-builder/src/main/java/datadog/trace/agent/tooling/AgentInstaller.java [hunk 2]
```diff
@@ -125,6 +125,24 @@ public static ClassFileTransformer installBytebuddyAgent(
...
+                  public void onTransformation(
+                      TypeDescription typeDescription,
+                      ClassLoader classLoader,
+                      JavaModule module,
+                      boolean loaded,
+                      DynamicType dynamicType) {
+                    try {
+                      dynamicType.saveIn(
+                          new java.io.File(
+                              "/home/malvarez/go/src/github.com/DataDog/dd-trace-java/build/instrumented"));
```
Comment (smola) [inline dd-java-agent/agent-builder/src/main/java/datadog/trace/agent/tooling/AgentInstaller.java:140]: I guess this is why this is still in draft :wink:

Comment (manuel-alvarez-alvarez) [reply dd-java-agent/agent-builder/src/main/java/datadog/trace/agent/tooling/AgentInstaller.java]: Yep, actually it's ready now :smile:

File: dd-java-agent/instrumentation/commons-httpclient-2/src/main/java/datadog/trace/instrumentation/commonshttpclient/IastCommonsHttpClientInstrumentation.java [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+
+import com.google.auto.service.AutoService;
+import datadog.trace.agent.tooling.Instrumenter;
+import datadog.trace.api.iast.InstrumentationBridge;
+import datadog.trace.api.iast.sink.SsrfModule;
+import net.bytebuddy.asm.Advice;
+import org.apache.commons.httpclient.HttpMethod;
+
+@AutoService(Instrumenter.class)
+public class IastCommonsHttpClientInstrumentation extends Instrumenter.Tracing
```
Comment (smola) [inline dd-java-agent/instrumentation/commons-httpclient-2/src/main/java/datadog/trace/instrumentation/commonshttpclient/IastCommonsHttpClientInstrumentation.java:16]:
```suggestion
public class IastCommonsHttpClientInstrumentation extends Instrumenter.Iast
```

**Review by smola — CHANGES_REQUESTED** [review]

File: dd-java-agent/instrumentation/commons-httpclient-2/src/main/java/datadog/trace/instrumentation/commonshttpclient/IastCommonsHttpClientInstrumentation.java (hunk 3 above)
Comment (manuel-alvarez-alvarez) [reply dd-java-agent/instrumentation/commons-httpclient-2/src/main/java/datadog/trace/instrumentation/commonshttpclient/IastCommonsHttpClientInstrumentation.java]: Fixed, thanks!

File: dd-java-agent/instrumentation/java-net/src/main/java/datadog/trace/instrumentation/java/net/URICallSite.java [hunk 4]
```diff
@@ -51,4 +51,33 @@ public static URI afterCtor(
     }
     return result;
   }
+
+  @CallSite.After("java.lang.String java.net.URI.toString()")
```
[the remark this answers is not in the document]
Comment (manuel-alvarez-alvarez) [reply dd-java-agent/instrumentation/java-net/src/main/java/datadog/trace/instrumentation/java/net/URICallSite.java:55]: The class has the annotation:

```java
@IastAdvice.Propagation(PropagationTypes.URI)
@CallSite(spi = IastAdvice.class)
public class URICallSite {
   ...
}
```

Method annotations take precedence over the type annotation so you can override the telemetry annotation.

**Review by smola — APPROVED** [review]

**Pull request merged**